### PR TITLE
feat: separate rendering and instance generation for the RubiksCube environment

### LINKED ISSUE #81: feat(rubiks): configurable generator and viewer

### PR DESCRIPTION
**Pull request by TristanKalloniatis**
Resolves #81

### CODE REVIEW

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/env_viewer.py:74]:
```suggestion
            cube_size: size of cube to view.
```

File: jumanji/environments/logic/rubiks_cube/generator.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/generator.py:30]:
```suggestion
    """Base class for generators for the `RubiksCube` environment."""
```

File: jumanji/environments/logic/rubiks_cube/generator.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/generator.py:33]:
```suggestion
        """Initialises a `RubiksCube` generator for resetting the environment.
```

File: jumanji/environments/logic/rubiks_cube/generator.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/generator.py:50]:
```suggestion
    """Generates instances by applying a given number of scrambles to a solved cube."""
```

File: jumanji/environments/logic/rubiks_cube/utils.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/utils.py:68]:
```suggestion
    """A sparse reward function: +1 if the cube is solved, otherwise 0."""
```

File: jumanji/environments/logic/rubiks_cube/utils.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/utils.py:121]: Would you mind fixing all the docstrings as done above?

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:235]: I copied it from another docstring (2048 I believe), otherwise no preference

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:258]: I agree but my personal preference is to be super explicit about named args

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: I could but it breaks the point of being able to configure the env viewer

Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: Btw in case it's not clear, DefaultRubiksViewer is the one that is actually implemented whereas RubiksCubeViewer is the abstract class

Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: Apologies, my question should have been why you can't put `DefaultRubiksViewer` in the type annotation for now. This would only be temporary since I'll create a PR with a viewer interface to be used across all environments. But it's probably best to do that PR before merging this one. Let me know if you agree/disagree

Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: My only argument against this (aside from wanting to close this PR asap lol) is that some env viewers might be a bit less generic than the base class you have in mind - eg mine has a cube_size. It's not strictly necessary but I figured that any viewer for Rubik's cube would have this at a minimum.

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env_viewer.py:26]: After #99 is merged, we can remove this abstract viewer and just keep one concrete viewer that we may want to call `RubiksCubeViewer`.

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env_viewer.py:68]: After #99 is merged, we can rename this
```suggestion
class RubiksCubeViewer(Viewer[State]):
```

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env.py:214]: Would we like to keep the notion of `action_history` (or rather sequence of actions used to scramble) in the `ScramblingGenerator` instead of removing it completely?

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Looks even cleaner now, thanks! I think we need to wait for #99 to be merged (@dluo96) to remove the abstract rubiks cube viewer.

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:214]: It is accessible by following through what the generator does, but I thought we agreed not to explicitly store it anywhere?

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env_viewer.py:26]: OK but then what would the type hint of the env viewer in the main env file be?

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: Wouldn't type hinting make it seem like you can pass another env's viewer when you really can't?

Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env.py:30]: @TristanKalloniatis it wouldn't - because each environment has a specific `State` - and this is specified in the type annotation `Viewer[State]`

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env_viewer.py:26]: It would be `Optional[Viewer[State]]`

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (clement-bonnet) [reply jumanji/environments/logic/rubiks_cube/env.py:230]: In most environments, we try to return whatever the viewer returns so that in case the viewer is one that outputs a numpy array we have it. Please see e.g. TSP render.

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (clement-bonnet) [reply jumanji/environments/logic/rubiks_cube/env.py:214]: You mean the sequence of actions used to reset is accessible by calling `generator.generate_actions_for_scramble(key)` on the same key used for reset? Then, I'm happy with this and we can resolve this.

Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py:214]: Correct

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env.py:88]: Could we call the environment viewer `viewer` to be consistent with the other environments?
```suggestion
        viewer: Optional[Viewer[State]] = None,
```

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env.py:89]: Would it be possible to reorder the init arguments by importance order, i.e. have generator first, then time_limit, reward_fn and viewer.

File: jumanji/environments/logic/rubiks_cube/env_viewer.py
Comment (clement-bonnet) [inline jumanji/environments/logic/rubiks_cube/env_viewer.py:27]: Should we rename this `RubiksCubeViewer` as it is obviously the default because the only implemented? 
```suggestion
class RubiksCubeViewer(Viewer[State]):
```

Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env_viewer.py]: Good idea

Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env_viewer.py]: This is also consistent with other environments

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (dluo96) [reply jumanji/environments/logic/rubiks_cube/env.py]: Agreed, similar to other environments

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (TristanKalloniatis) [reply jumanji/environments/logic/rubiks_cube/env.py]: (I changed it to be consistent)

File: jumanji/environments/logic/rubiks_cube/env.py
Comment (dluo96) [inline jumanji/environments/logic/rubiks_cube/env.py:104]:
```suggestion
                Implemented options are [`RubiksCubeViewer`]. Defaults to `RubiksCubeViewer`.
```

**Pull request merged**
